src/str/transform.php

```diff
@@ -242,9 +244,96 @@ function replace_every_ci(
   );
 }
 
-function reverse(
-  string $string,
+/**
+ * Returns the "haystack" string with all occurrences of the keys of
+ * `$replacements` replaced by the corresponding values.
+ *
+ * Once a substring has been replaced, its new value will not be searched
+ * again.
+ *
+ * - For having new values searched again, see `Str\replace_every()`.
+ */
+<<__Rx>>
+function replace_every_nonrecursive(
+  string $haystack,
+  KeyedContainer<string, string> $replacements,
+): string {
+  invariant(
+    !C\contains_key($replacements, ''),
+    'Expected non-empty keys only.',
+  );
+
+  /* HH_FIXME[2049] calling stdlib directly */
+  /* HH_FIXME[4107] calling stdlib directly */
+  /* HH_FIXME[4200] Rx calling non-rx */
+  return \strtr($haystack, $replacements);
+}
+
+/**
+ * Returns the "haystack" string with all occurrences of the keys of
+ * `$replacements` replaced by the corresponding values (case-insensitive).
+ *
+ * Once a substring has been replaced, its new value will not be searched
+ * again.
+ * If a replacer is the prefix of another (like "car" and "carpet"), the longer
+ * one (carpet) gets replaced, if there is any ambiguity.
+ * When two replacers are passed that are identical except for case
+ * an invariant exception is thrown.
+ *
+ * - For having new values searched again, see `Str\replace_every_ci()`.
+ *
+ * Time complexity O(length * a) where `length` is the length of the subject
+ * and a is the amount of replacers with different lengths.
+ */
+<<__Rx>>
+function replace_every_nonrecursive_ci(
+  string $haystack,
+  KeyedContainer<string, string> $replacements,
 ): string {
+  invariant(
+    !C\contains_key($replacements, ''),
+    'Expected non-empty keys only.',
+  );
+
+  $haystack_lc = lowercase($haystack);
+  $key_lengths = keyset[];
+  $replacements_lc = dict[];
+  foreach ($replacements as $key => $value) {
+    $key_lc = lowercase($key);
+    invariant(
+      !C\contains_key($replacements_lc, $key_lc),
+      'Duplicate case-insensitive search string "%s".',
+      $key_lc,
+    );
+    $key_lengths[] = length($key_lc);
+    $replacements_lc[$key_lc] = $value;
+  }
+
+  $key_lengths = Vec\sort($key_lengths) |> Vec\reverse($$);
+
+  $output = '';
+  for ($pos = 0; $pos < length($haystack); ) {
+    $found_match_at_pos = false;
+    foreach ($key_lengths as $key_length) {
+      $possible_match = slice($haystack_lc, $pos, $key_length);
+      if (C\contains_key($replacements_lc, $possible_match)) {
+        $found_match_at_pos = true;
+        $output .= $replacements_lc[$possible_match];
+        $pos += $key_length;
+        break;
+      }
+    }
+
+    if (!$found_match_at_pos) {
+      $output .= $haystack[$pos];
+      $pos++;
+    }
+  }
```

Comment:
The whole outer loop is therefore O(haystack length * sum of all distinct key lengths).